### ISSUE TITLE
[datasets] groupby random behaviour and dataset type

### ISSUE THREAD

**Issue by giucesar** [ray-project/ray#26286]
### What happened + What you expected to happen

Hi.
I am running a simple groupby in a ray.data.Dataset.

The dataset is a list of namedtuples (ex. NamedTuple("dummy", [("value_a", list), ("value_b", int)])).
When I run the group by, the key is one of the fields of the named tuple.

```
ds.groupby(lambda nt: nt.value_b).aggregate(AggregateFn(do stuff in a complicated way))

ds.groupby(lambda nt: nt.value_b).map_groups(do stuff in a simple way)
```

When I run it, sometimes it runs normally sometimes it complains about the key:
`Table sort key must be a column name, was: <function foo.<locals>.<lambda> at 0x10be2d1f0>`

From what I understood, when ds is an arrow table, it fails, when it is a simple list it works.

The puzzling part is that I can't figure out why randomly the same code generates arrow tables and lists when the schema never changes, they are always lists of namedtuples of the exact same type, just random values inside.

Is there any way to ensure the output of the dataset after a map_groups or map_batches is of the type I want? Also, is there any way to convert from one type to another? I tried all hidden methods but couldn't figure them out, also, the batch_format options don't help.

For the group by error, I am just following what the documentation suggests, but it fails (although I understand why). I believe there are some constraints on the way the groupby is defined. Is there any way to work around this?

### Versions / Dependencies

ray 1.13.0
python 3.9.12
macos 12.4

### Reproduction script

not necessarily a bug, maybe a clarification of documentation

### Issue Severity

Low: It annoys or frustrates me.

**Comment by xwjiang2010** [comment]
@clarkzinzow @jianoaix

**Comment by clarkzinzow** [comment]
@giucesar Thank you for filing the issue!

That error should only pop up when trying to supply a lambda groupby key for a tabular dataset, since this isn't supported. Can you share more of how the dataset `ds` is created?

**Comment by giucesar** [comment]
For sure.
Initially, it is create like:
`ds = ray.data.range(n_tuples).map_batches(lambda batch: generate_random_tuples(size=len(batch))`

After some processing, it gets to a point where I need to transform the data into groups defined by one of the values in the named tuples.
`ds = ds.group_by(lambda row: row.key).map_groups(lambda group: process_group_and_return_same_size_and_type(group))`

When the group by executes, it randomly returns a simple dataset or a pyarrow dataset. I can't figure out why. Probably something related to the distribution of the blocks, hard to say.

**Comment by clarkzinzow** [comment]
@giucesar I tried to reproduce this issue following a similar flow on latest master but I wasn't able to encounter such a failure, could you post a reproduction script here the triggers this failure?

**Comment by giucesar** [comment]
It still happens randomly.
I tried to write a smaller script for you to reproduce but failed. From what I figured out, the named tuple has two dictonaries with named tuple values (dict[str, NamedTuple]) on it. I tend to believe that the overcomplicated data structure is the cause of it. 

`NamedTuple('dummy',[('name',str),('table_a',NamedTuple('tablea',[('val1',int)...]),('table_b',NamedTuple('tableb',[('valt',int)...])])`

With that, I rewrote the code to be simpler and it is working consistently now.
If this edge case shows up again, I'll try to isolate it.
Sorry for the trouble.
